### PR TITLE
feat(docker): add Dockerfile and Docker compose setup

### PR DESCRIPTION
**Pull request by fh1ch**
This change adds the posibilitiy to build a Docker Image out of Fossology.
It allows the execution as standalone container (combined database) or
connection against an external database instance as used by the Docker compose setup.

Signed-off-by: Roger Meier <r.meier@siemens.com>
Signed-off-by: Fabio Huser <fabio.huser@siemens.com>
Tested-by: Maximilian Huber <maximilian.huber@tngtech.com>
Tested-by: Pascal Bach <pascal.bach@siemens.com>

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
Comment (mcjaeger) [inline docker-entrypoint.sh]: great work, may I ask for a super-small addition? .. could you pls. add headers to the files created (maybe also the docker file) with permission notice and disclaimer:

# Copying and distribution of this file, with or without modification,
# are permitted in any medium without royalty provided the copyright
# notice and this notice are preserved.  This file is offered as-is,
# without any warranty.

like here (scroll down):
https://github.com/fossology/fossology/wiki/Coding-Style

**Comment by mcjaeger** [comment]
about the adding of the header: it makes clearing a lot easier - I understand that you have a super clean ("vorbildlich") pull request: clean git message, clean sign-off, clean testing. Let's adopt it for other pull requests in future.

File: docker-entrypoint.sh (hunk 1 above)
Comment (fh1ch) [inline docker-entrypoint.sh]: @mcjaeger Thanks for the review!

Sure, I've just added the file header you've mentioned to the `Dockerfile`, `docker-compose.yml` file and the `docker-entrypoint.sh` startup script. They're all squashed into the same commit.

**Comment by maximilianhuber** [comment]
This branch is >322 commits behind master and the version on docker hub is 2.6.1. I think you will create an updated image for docker hub once this is merged?

**Comment by maximilianhuber** [comment]
When I call `docker-compose up` I see the following in the log
tmp
this error seems to come from 
<pre>
<b>install/db/dbcreate.in:15:</b>   echo "ERROR: postgresql isn't running"
</pre>
could it be that the DB is not starting fast enougth (althoug the log states immediately above this, that it is ready. The `psql` call might have been issued before this.)

**Comment by bufferoverflow** [comment]
I will add a hook similar to travis.yml so we will have a ready to use image with each push to master and versioned Docker Images for tags.

**Comment by maximilianhuber** [comment]
@bufferoverflow: Ok, that is a great solution.

**Comment by maximilianhuber** [comment]
Except for the problem with `install/db/dbcreate.in` everything looks fine. I will merge this PR once the problem mentioned above is discussed / resolved.

**Comment by bufferoverflow** [comment]
What about this?
```
diff --git a/install/db/dbcreate.in b/install/db/dbcreate.in
index 6efe349..b044f0c 100644
--- a/install/db/dbcreate.in
+++ b/install/db/dbcreate.in
@@ -5,6 +5,12 @@
 # This script checks to see if the the fossology db exists and if not
 # then creates it.
 
+# exit when using an external DB
+if [ "$FOSSOLOGY_DB_HOST" ]
+  echo $FOSSOLOGY_DB_HOST
+  exit 0
+fi
+
 echo "*** Setting up the FOSSology database ***"
 
 # At some point this is where we could dynamically set the db password
```

**Comment by maximilianhuber** [comment]
Is this script not necessary for an external DB? Then this patch would solve the problem.

For the test one should add `-n` or something simiar.
Maybe a more verbose output would be good.

**Comment by bufferoverflow** [comment]
I guess we need parts of `install/db/fossologyinit.sql`, especially the `SET client_encoding = 'SQL_ASCII';` that is set within this sql script.

using the update patch above fixes the symptom, but I still have the following issues
```
db_1  | ERROR:  relation "spdx2_ars" already exists
db_1  | STATEMENT:  CREATE TABLE spdx2_ars() INHERITS(ars_master);
db_1  | 	    ALTER TABLE ONLY spdx2_ars ADD CONSTRAINT spdx2_ars_agent_fk_fkc FOREIGN KEY (agent_fk) REFERENCES agent(agent_pk);
db_1  | 	    ALTER TABLE ONLY spdx2_ars ADD CONSTRAINT spdx2_ars_upload_fk_fkc FOREIGN KEY (upload_fk) REFERENCES upload(upload_pk) ON DELETE CASCADE
db_1  | ERROR:  relation "spdx2_ars" already exists
db_1  | STATEMENT:  CREATE TABLE spdx2_ars() INHERITS(ars_master);
db_1  | 	    ALTER TABLE ONLY spdx2_ars ADD CONSTRAINT spdx2_ars_agent_fk_fkc FOREIGN KEY (agent_fk) REFERENCES agent(agent_pk);
db_1  | 	    ALTER TABLE ONLY spdx2_ars ADD CONSTRAINT spdx2_ars_upload_fk_fkc FOREIGN KEY (upload_fk) REFERENCES upload(upload_pk) ON DELETE CASCADE
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
db_1  | ERROR:  invalid byte sequence for encoding "UTF8": 0xa9
db_1  | STATEMENT:  INSERT INTO copyright(agent_fk, pfile_fk, content, hash, type, copy_startbyte, copy_endbyte) VALUES($1,$2,$3,md5($3),$4,$5,$6)
```

**Comment by bufferoverflow** [comment]
initialize the db with `SQL_ASCII` charset within docker-compose fixes related UTF-8 issue above.
```
diff --git a/install/docker-compose.yml b/install/docker-compose.yml
index b5c90af..3b250ac 100644
--- a/install/docker-compose.yml
+++ b/install/docker-compose.yml
@@ -26,3 +26,4 @@ db:
     - POSTGRES_DB=fossology
     - POSTGRES_USER=fossy
     - POSTGRES_PASSWORD=fossy
+    - POSTGRES_INITDB_ARGS='-E SQL_ASCII'
```
I will update the commit with these changes.

**Comment by maximilianhuber** [comment]
As an alternative you could let docker load the file `fossologyinit.sql` via
```
ADD install/db/fossologyinit.sql /docker-entrypoint-initdb.d/
```

The current postgres image automatically runs `.sql` scripts in this folder (see  https://hub.docker.com/_/postgres/).

This also substitutes the lines
```
     - POSTGRES_DB=fossology
     - POSTGRES_USER=fossy
     - POSTGRES_PASSWORD=fossy
```

**Comment by bufferoverflow** [comment]
Yes, but this would require a specific `fossology/fossology-postgresql` image and an additional Dockerfile to handle this. I personally prefer to use the upstream version as it is.

**Comment by maximilianhuber** [comment]
Why are you able to ignore files such that `fossologyinit.sql`, aren't they essential for the installation/setup?

Can they be removed from the other installation variants as well?

**Comment by maximilianhuber** [comment]
There are problems with .deb-files:
On a fresh generated container I was not able to extract debian files. The upload is not even clickeable in the browse view:
tmp5

In Vagrant it works (under ubuntu instead of debian).
Does someone else also have this issue?

**Comment by maximilianhuber** [comment]
There are problems with the copyright and ecc agents:
On a rather fresh install the agents **copyright** and **ecc** do not start.

tmp5

Does someone else also have this issue?

(I will rebuild and test again)

**Comment by bufferoverflow** [comment]
copyright and ecc is ok with latest commit here using zlib.

I will dig into the deb unpack issue right now and study `fossologyinit.sql` in detail. Did you had some encoding issues in the past? Any reason or idea why  SQL_ASCII instead of UTF-8 was chosen as encoding?

**Comment by maximilianhuber** [comment]
I do not know why SQL_ASCII is chosen, maybe @steffen-weber could help.

**Comment by maximilianhuber** [comment]
The **copyright** and **ecc** ecc problem persists for me.

What was done:
* remove all docker container and images
* `git fetch siemens`
* `git reset hard siemens/feat/docker` (but set the http port for fossology to 8082, 8081 is used)
* `cd install && docker-compose up`
* upload zlib  with **Upload/From File** and all agents enabled

then:
tmp6

**Comment by bufferoverflow** [comment]
Did you delete all existing docker container's (`docker images`, `docker ps -a`)?
I made a fresh `docker-compose up` based on f384b2bad9a1e4ac9bbb62606f551b79942eb8fd

ecc and copyright is ok with zlib but deb packages is still broken

I guess `fossologyinit.sql` needs a visit ;-)

**Comment by maximilianhuber** [comment]
For cleanup I have used
```
docker ps -aq |xargs docker rm
docker images -aq |xargs docker rmi
```
after that, neither `docker ps -a` nor `docker images -a` lists anything.

**Comment by maximilianhuber** [comment]
I am currently testing it using vagrant and the following Vagrantfile:
```
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure(2) do |config|
  config.vm.box = "box-cutter/ubuntu1404-docker"
  config.vm.provider "virtualbox" do |vb|
  #   vb.gui = true
    vb.memory = "6000"
  end
  config.vm.network "forwarded_port", guest: 8081, host: 8081
  config.vm.provision "shell", inline: <<-SHELL
    cd /vagrant/
    sudo apt-get update
    sudo apt-get install -y git
    [ -e docker-compose ] || curl -L https://github.com/docker/compose/releases/download/1.7.0/docker-compose-`uname -s`-`uname -m` > docker-compose 2> /dev/null
    chmod +x docker-compose
    [ -e fossology ] || git clone -b feat/docker --single-branch https://github.com/siemens/fossology
    cd fossology/install
    ../../docker-compose build | tee ../../docker-compose_build.log
    (../../docker-compose up | tee ../../docker-compose_up.log) &
  SHELL
end
```

There I have the problem that the db gets ready after the web, i.e.:
tmp

**Comment by maximilianhuber** [comment]
There is always also the following message in the log:
tmp
I am not shure whether this is a debian, fossology or docker problem. One should first checkt, whether this also appears in the old vagrant installation.

**Comment by maximilianhuber** [comment]
A solution for the timing problem is
```
diff --git a/docker-entrypoint.sh b/docker-entrypoint.sh
index 61bf88e..d4fa10d 100644
--- a/docker-entrypoint.sh
+++ b/docker-entrypoint.sh
@@ -27,6 +27,15 @@ if [ "$FOSSOLOGY_DB_PASSWORD" ]; then
        db_password="$FOSSOLOGY_DB_PASSWORD"
 fi
 
+testForPostgres(){
+    PGPASSWORD=$db_password psql -h "$db_host" "$db_name" "$db_user" -c '\l' >/dev/null
+    return $?
+}
+until testForPostgres; do
+    >&2 echo "Postgres is unavailable - sleeping"
+    sleep 1
+done
+
 # Write configuration
 cat <<EOM > /usr/local/etc/fossology/Db.conf
 dbname=$db_name;
```

See PR https://github.com/siemens/fossology/pull/7.

**Pull request merged**